### ISSUE TITLE
accept raster / array of pour points as input to catchment

### ISSUE THREAD

**Issue by smnorris** [mdbartos/pysheds#64]
My use case is for delineating small catchments, forcing the catchments to match existing streams. 
I can fill/inflate the DEM, burn stream lines into the DEM, and snap pour points to high accumulation cells. This works very well for most sites.

The exception is sites where the DEM isn't quite clean enough - there is a bump somewhere along a known stream so some catchment area is missed. Simply expanding the stream lines burned into the DEM didn't help (`stream_raster = maximum_filter(stream_raster, 2)`). Filtering or rounding the DEM is worth trying, or I could even burn the z-values of my known streams into the DEM. But I think the simplest way to ensure all required catchment area is captured for my use case is to: 
1. generate multiple pour points along the stream
2. run the catchment function at each point
3. aggregate the results

I'll do this - but wonder if it would be worth tweaking the catchment function to also accept pour points as either a raster / array or list of points?

**Comment by smnorris** [comment]
Pretty much.  Ideally I want `catchment` output to be contributing area of all non-zeros or no-nodata in an input 'pour point' raster.  This reduces need for cleaning DEM if I have good stream data already.

See `in_pour_point_data` parameter [here](http://desktop.arcgis.com/en/arcmap/latest/tools/spatial-analyst-toolbox/watershed.htm) 
```
For a raster, this represents cells above which the contributing area, or catchment, will be determined. 
```

**Comment by smnorris** [comment]
Turns out that just adding one line fixes my issue for all test sites.

```
# Set mask to blurred mask
mask = blurred_mask
# Create a view onto the DEM array
dem = grid.view('dem', dtype=np.float64, nodata=np.nan)
# Subtract dz where mask is nonzero
dem[(mask > 0)] -= dz*mask[(mask > 0)]
# Smooth the DEM in masked area - trying to force the DEM to go downhill
# along the stream
dem[(mask > 0)] = ndimage.filters.minimum_filter(dem, 2)[(mask > 0)]
```

**Comment by mdbartos** [comment]
You've tried `grid.fill_depressions` followed by `grid.resolve_flats` right? That should force everything to route to the boundaries of the dataset.

**Comment by mdbartos** [comment]
Hi @smnorris 

If you want to send the data to me by email (see profile), I can take a look at it. Just let me know what problem(s) you are still having and need to resolve.

Thanks,
MDB

**Comment by smnorris** [comment]
Here are a handful of DEMs where I'm having the problem, plus the stream segments I'm burning in (and would have used as pour point rasters with the ArcGIS method)
https://hillcrestgeo.ca/outgoing/public

This shows what I'm doing, using the first DEM:
https://gist.github.com/39efa810e510acf418268c2438abeb68

Here is a view of the inputs and outputs in QGIS (point is the pour point, input rasterized stream segment in red).  The streams shown on the bottom half of the image all flow to the pour point. ![screenshot 2019-02-04 14 30 13](https://user-images.githubusercontent.com/1323293/52241760-67e1e980-2889-11e9-98dc-b8516f0d1136.png)

**Comment by cfandel** [comment]
Hi, 

I actually would like the opposite functionality. I have a bunch of points, and I'd like to be able to pass lists of x and y coordinates for those points to grid.catchment, but get the catchment _for each point_, not for all the points together - similar to what the ArcMap Spatial Analyst Watershed tool does if you give it multiple pour points. The lower points should have catchments that exclude areas already captured by the higher points. Is this possible?

Should look something like this:
![image](https://user-images.githubusercontent.com/31896726/52650138-f25dc500-2ee9-11e9-894b-dc2daac37d3f.png)

**Comment by mdbartos** [comment]
Hi @cfandel, thanks for your question. You can do this using a `for` loop. If your points are not topologically ordered, you can do something like this:

```python
# Import modules
import numpy as np
import matplotlib.pyplot as plt
from pysheds.grid import Grid
import seaborn as sns
np.random.seed(0)

# Load flow direction grid
grid = Grid.from_raster('../data/n30w100_dir', data_name='dir')

# Generate the small sample catchment
x, y = -97.294167, 32.73750
grid.catchment(data='dir', x=x, y=y, out_name='catch',
               recursionlimit=15000, xytype='label', nodata_out=0)
grid.clip_to('catch')

# Get some example pour points
grid.accumulation(data='dir', out_name='acc', apply_mask=False)
branches = grid.extract_river_network('catch', 'acc')
points = np.vstack([geom['geometry']['coordinates'][-1] for geom in branches['features']])
points = points[np.random.choice(len(points), size=10, replace=False), :]
# Add outlet point
points = np.vstack([points, np.asarray([x, y])])

# Delineate catchments
labels = np.zeros(grid.shape, dtype=int)
catchments = []
for point in points:
    x, y = point
    catchment = grid.catchment(x=x, y=y, data='dir', xytype='label', inplace=False)
    catchments.append(catchment)
    labels += (catchment != 0)

# Find non-overlapping catchments
for index, catchment in enumerate(catchments):
    mask = (catchment > 0)
    label = labels[mask].min()
    catchments[index] = np.where(mask & (labels == label), catchment, 0)

labeled_catchments = sum([np.where(catchment, index + 1, 0) 
                          for index, catchment in enumerate(catchments)]).astype(float)
labeled_catchments[labeled_catchments == 0] = np.nan

# Plot results
fig, ax = plt.subplots(figsize=(8,6))
plt.imshow(labeled_catchments, cmap='cubehelix', zorder=1, extent=grid.extent)
plt.colorbar()
plt.scatter(points[:, 0], points[:, 1], zorder=2, c='r')
```

![multicatch](https://user-images.githubusercontent.com/8175317/52670413-73b35880-2ee6-11e9-95d2-c6c0166f7023.png)

**Comment by mdbartos** [comment]
Because there are several possible interpretations for what result an array of pour points should generate, I would prefer not to overload the inputs to the `catchment` function. However, I am interested in creating another function that is analogous to the `watershed` functions from scikit-image or scipy.ndimage:

http://scikit-image.org/docs/dev/auto_examples/segmentation/plot_watershed.html
http://scipy-lectures.org/advanced/image_processing/auto_examples/plot_watershed_segmentation.html

In other words, supply an array of labels and delineate the contributing area for each labeled region.

**Comment by cfandel** [comment]
This is perfect thank you! I was getting stuck on how to get the subcatchments to not overlap. Do you mind if I take the code chunk above and modify/incorporate it into what I'm working on?

**Comment by mdbartos** [comment]
Sure! Go ahead. There is probably a more memory-efficient method if you can guarantee that the points are sorted (i.e. a point occurs later in the list if it is downstream of another point).

**Comment by cfandel** [comment]
They're definitely not going to be sorted, unfortunately (they're karst conduit inlet nodes), but this should be efficient enough. Thanks so much!

**Comment by newdeskarl** [comment]
Hi Matt Bartos, 

Congratulations and many thanks for this tool that I begin to use. 

Sorry to dig up the case. I encounter an issue in the code you shared. You started by uploading flow direction data from a raster file. But, me, I started by uploading my DEM data. So, my flow direction raster was calculated and stoked in a variable (fdir = grid.flowdir(inflated_dem, dirmap=dirmap)). I have trouble using it as a variable. Can you help me. Thanks.

THE ENTIRE CODE: 

import numpy as np
import matplotlib.pyplot as plt
from matplotlib import colors
import seaborn as sns
from pysheds.grid import Grid 
import fiona 
np.random.seed(0)

grid = Grid.from_raster('southernpeninsula_mnt_aster_v2_subset1_utm.tif')
dem = grid.read_raster('southernpeninsula_mnt_aster_v2_subset1_utm.tif')

fig, ax = plt.subplots(figsize=(8,6))
fig.patch.set_alpha(0)

plt.imshow(dem, extent=grid.extent, cmap='terrain', zorder=1)
plt.colorbar(label='Elevation (m)')
plt.grid(zorder=0)
plt.title('Digital elevation map', size=14)
plt.xlabel('Longitude')
plt.ylabel('Latitude')
plt.tight_layout()

# Fill pits in DEM
pit_filled_dem = grid.fill_pits(dem)

# Fill depressions in DEM
flooded_dem = grid.fill_depressions(pit_filled_dem)
    
# Resolve flats in DEM
inflated_dem = grid.resolve_flats(flooded_dem)

# Determine D8 flow directions from DEM

# Specify directional mapping
dirmap = (64, 128, 1, 2, 4, 8, 16, 32)
    
# Compute flow directions

fdir = grid.flowdir(inflated_dem, dirmap=dirmap)

fig = plt.figure(figsize=(8,6))
fig.patch.set_alpha(0)

plt.imshow(fdir, extent=grid.extent, cmap='viridis', zorder=2)
boundaries = ([0] + sorted(list(dirmap)))
plt.colorbar(boundaries= boundaries,
             values=sorted(dirmap))
plt.xlabel('Longitude')
plt.ylabel('Latitude')
plt.title('Flow direction grid', size=14)
plt.grid(zorder=-1)
plt.tight_layout() 

# Load flow direction grid
grid = grid.read_raster(fdir)

# Generate the small sample catchment
x, y = 594479.32, 2060013.15  # UTM coordinate
grid.catchment(data='dir', x=x, y=y, out_name='catch',
               recursionlimit=15000, xytype='label', nodata_out=0)
grid.clip_to('catch')

points = np.loadtxt('southernpeninsula_outlet.txt', skiprows=2) # file with coordinate list of pour points

# Get some example pour points
grid.accumulation(data='dir', out_name='acc', apply_mask=False)
branches = grid.extract_river_network('catch', 'acc')
points = np.vstack([geom['geometry']['coordinates'][-1] for geom in branches['features']])
points = points[np.random.choice(len(points), size=10, replace=False), :]
# Add outlet point
points = np.vstack([points, np.asarray([x, y])])

# Delineate catchments
labels = np.zeros(grid.shape, dtype=int)
catchments = []
for point in points:
    x, y = point
    catchment = grid.catchment(x=x, y=y, data='dir', fdir=grid, xytype='label', inplace=False)
    catchments.append(catchment)
    labels += (catchment != 0)

# Find non-overlapping catchments
for index, catchment in enumerate(catchments):
    mask = (catchment > 0)
    label = labels[mask].min()
    catchments[index] = np.where(mask & (labels == label), catchment, 0)

labeled_catchments = sum([np.where(catchment, index + 1, 0) 
                          for index, catchment in enumerate(catchments)]).astype(float)
labeled_catchments[labeled_catchments == 0] = np.nan

# Plot results
fig, ax = plt.subplots(figsize=(8,6))
plt.imshow(labeled_catchments, cmap='cubehelix', zorder=1, extent=grid.extent)
plt.colorbar()
plt.scatter(points[:, 0], points[:, 1], zorder=2, c='r') 

THE ERROR: 

Traceback (most recent call last):

  File ~/opt/anaconda3/lib/python3.9/site-packages/spyder_kernels/py3compat.py:356 in compat_exec
    exec(code, globals, locals)

  File ~/Documents/Haiti_2021EQ/southernpeninsula_pysheds.py:79
    grid = grid.read_raster(fdir)

  File ~/opt/anaconda3/lib/python3.9/site-packages/pysheds/sgrid.py:261 in read_raster
    return pysheds.io.read_raster(data=data, band=band, window=window,

  File ~/opt/anaconda3/lib/python3.9/site-packages/pysheds/io.py:99 in read_raster
    with rasterio.open(data, **kwargs) as f:

  File ~/opt/anaconda3/lib/python3.9/site-packages/rasterio/env.py:451 in wrapper
    return f(*args, **kwds)

  File ~/opt/anaconda3/lib/python3.9/site-packages/rasterio/__init__.py:216 in open
    raise TypeError("invalid path or file: {0!r}".format(fp))

TypeError: invalid path or file: Raster([[ 0,  0,  0, ...,  0,  0,  0],
        [ 0, 16, 64, ..., 64,  1,  0],
        [ 0, 16, 16, ...,  1,  1,  0],
        ...,
        [ 0, 16, 16, ...,  4,  1,  0],
        [ 0, 16,  4, ...,  4,  1,  0],
        [ 0,  0,  0, ...,  0,  0,  0]])
